[Chan Sze Ann Eugene] IP

eugenecsa opened the pull request:
# Doke

> S"low down and enjoy life. It's not only the scenery you miss by going too fast - you also miss the sense of where you are going and why." - Eddie Cantor

Doke frees your mind of having to remember things you don't need to. It's
- text-based
- easy to learn
- **SUPER** FAST & RESPONSIVE

All you need to do is,
1. download it from [here](https://github.com/eugenecsa/ip/releases/tag/v0.1)
2. double-click it
3. add your tasks
4. let it manage your task for you!! :100: :100:

And it's all **FREE!!!**

Features:
- [x] Managing tasks
- [x] Managing deadlines (~~coming soon~~)
- [ ] Reminder (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.luanch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

eugenecsa committed:
Edited Duke.java file for Level-1 (Greet, Echo, Exit).

eugenecsa committed:
Modified Duke.java and added new file List.java for Level-2

eugenecsa committed:
Modified Duke.java and List.java, added Task.java for Level-3.

eugenecsa committed:
Modified Duke.java and List.java. Added Deadline.java, Event.java, ToDo.java. Level-4.

eugenecsa committed:
Modified input and expected file for I/O redirection testing.

eugenecsa committed:
Added exception handling for some of the inputs.

eugenecsa committed:
Added functionality for deletion of tasks.

eugenecsa committed:
Add functionality to save and load tasks to and from file.

eugenecsa committed:
Change deadline from String to LocalDate objects.

eugenecsa committed:
Merge branch 'branch-Level-7'

eugenecsa committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java
#	src/main/java/List.java

eugenecsa committed:
Refactor code and extract out classes

eugenecsa committed:
Add classes into package.

eugenecsa committed:
Add JUnit test for TaskList class.

eugenecsa committed:
Package the App as a JAR file.

eugenecsa committed:
Add JavaDoc comments to the code.

eugenecsa committed:
Edit code to comply with coding standard.

eugenecsa committed:
Add functionality to find tasks based on keyword.

Imranr2 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,129 @@
+     * @param input The input String.
+     * @return Boolean value indicating if program should continue.
+     */
+    public boolean process(String input) {
```
Should the different cases be extracted out as separate methods?
https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods

> Be wary when a method is longer than the computer screen, and take corrective action when it goes beyond 30 LOC (lines of code). The bigger the haystack, the harder it is to find a needle.

jodichoo replied:
Should this method be named like 'shouldContinue' since it is a boolean method?

Imranr2 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,174 @@
+        }
+    }
+
+    public ArrayList<String> findTask(String keyword) {
```
Perhaps you could also add JavaDocs for this public method?

Imranr2 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,174 @@
+     *
+     * @param description The description of the task.
+     */
+    public String addTask(String description) {
```
Is there a reason why you decided to name these methods all addTask rather than addToDo, addDeadline, addEvent?

Imranr2 commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,174 @@
+     * @param taskNumber The index(plus 1) of the task to be marked as done.
+     * @return The string representation of the task after it is marked as done.
+     */
+    public String changeTaskStatus(int taskNumber) throws DukeException {
```
Do you think it would be better to extract out the checking of the index for both the methods changeTaskStatus and deleteTask into a separate method instead?

Imranr2 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,58 @@
+     * @return String representation of the task for file writing.
+     */
+    @Override
+    public String saveString() {
```
I like how you implemented the general method in the Task class and override the implementation in its subclasses.

Imranr2 reviewed (commented):
Really well done! Your code follows the coding standard very closely. Just some nits to fix

jodichoo commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,102 @@
+     * @return ArrayList containing string representation of the tasks.
+     * @throws DukeException No current save file.
+     */
+    public ArrayList<String> load() throws DukeException {
```
Perhaps you could name this method loadTasks, since it returns a collection of tasks?

jodichoo commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,174 @@
+                addTask(splitData[2], LocalDate.parse(splitData[3]),"deadline", Boolean.parseBoolean(splitData[1]));
+                break;
+            case("E"):
+                addTask(splitData[2], LocalDate.parse(splitData[3]), "event", Boolean.parseBoolean(splitData[1]));
```
Lines 39 and 42 appears to have exceeded the soft limit on line length (110 chars), perhaps line breaks can help increase the readability for these parts!

jodichoo commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,174 @@
+     *
+     * @param description The description of the task.
+     * @param date The date/time attached to the task.
+     * @param type Indicates if the task is a Deadline or Event.
```
Is the Javadoc comment for the return value missing?

jodichoo commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,10 @@
+ * Represents an exception that occurs when running the Duke program. Takes in an error message.
+ */
+public class DukeException extends Exception{
+    public DukeException(String message) {
```
Is this missing a Javadoc comment?

jodichoo commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,129 @@
+     * @return Boolean value indicating if program should continue.
+     */
+    public boolean process(String input) {
+        boolean end = false;
```
The same issue with boolean variables as pointed out in the previous comment 😅

jodichoo reviewed (commented):
Hello! In general your code followed the java coding standards pretty strictly, aside from a few minor violations I noticed. Your code was really readable too 👍

eugenecsa committed:
Merge remote-tracking branch 'origin/add-gradle-support' into branch-A-Gradle

eugenecsa committed:
Add support for Gradle to the project.

eugenecsa committed:
Merge branch 'branch-A-Gradle'

eugenecsa committed:
Add GUI support for Duke chatbot and remove text input.

eugenecsa committed:
Merge branch 'branch-Level-10'

eugenecsa committed:
Add assertions to the project.

The code does not have any assertions to check for bugs in the code.

Using assert feature (not JUnit assertions) to document important
assumptions that should hold at various points in the code can catch
possible bugs.

Assertions are used to define assumptions about the program state so
that the runtime can verify them. An assertion failure indicates a
possible bug in the code because the code has resulted in a program
state that violates an assumption about how the code should behave.

eugenecsa committed:
Improve code quality

There are some code that does not adhere to guidelines under code quality.

Changing code to adhere to code quality improves readability and is easier to debug and understand.

Let's improve the code quality by refactoring the code and following the code quality guidelines given.

Commit:
Merge pull request #1 from eugenecsa/branch-A-CodeQuality

Improve code quality

eugenecsa committed:
Merge branch 'master' of https://github.com/eugenecsa/ip

* 'master' of https://github.com/eugenecsa/ip:
  Improve code quality

eugenecsa committed:
Merge branch 'master' into branch-A-Assertions

# Conflicts:
#	src/main/java/duke/TaskList.java

Commit:
Merge pull request #2 from eugenecsa/branch-A-Assertions

Add assertions to the project.

eugenecsa committed:
Merge branch 'master' of https://github.com/eugenecsa/ip

* 'master' of https://github.com/eugenecsa/ip:
  Add assertions to the project.

eugenecsa committed:
Add support for tagging tasks

Tasks does not support tagging.

Tagging tasks allows grouping of tasks with tags, and searching for a tag will return all tasks with that tag.

Let's,
Add a field in the Task class to contain the tag.
Add support for tagging the tasks.
Add Support for searching of tasks by a tag.

eugenecsa committed:
Merge branch 'branch-C-Tagging'

eugenecsa committed:
Update GUI

eugenecsa committed:
Add Ui.png

eugenecsa committed:
Update README.md

Update README.md to contain the user guide for Duke.

Commit:
Update README.md

eugenecsa committed:
Modify file path

eugenecsa committed:
Merge branch 'master' of https://github.com/eugenecsa/ip

eugenecsa committed:
Edit README.md

Commit:
Update README.md

Commit:
Update README.md
